src/libraries/System.Private.CoreLib/src/System/Threading/Thread.Unix.cs

```diff
@@ -5,22 +5,34 @@
 using System.Runtime;
 using System.Runtime.InteropServices;
 using Microsoft.Win32.SafeHandles;
+using System.Threading.Tasks;
 
 namespace System.Threading
 {
     public sealed partial class Thread
     {
         // these methods are temporarily accessed via UnsafeAccessor from generated code until we have it in public API, probably in WASI preview3 and promises
 #if TARGET_WASI
-        internal static System.Threading.Tasks.Task RegisterWasiPollable(int handle)
+        internal static System.Threading.Tasks.Task RegisterWasiPollableHandle(int handle)
         {
-            return WasiEventLoop.RegisterWasiPollable(handle);
+            return WasiEventLoop.RegisterWasiPollableHandle(handle);
         }
 
-        internal static void DispatchWasiEventLoop()
+        internal static int PollWasiEventLoopUntilResolved(Task<int> mainTask)
         {
-            WasiEventLoop.DispatchWasiEventLoop();
+            while (!mainTask.IsCompleted)
```

Comment:
This is probably too naive solution. 
There could be other pending tasks scheduled to C# thread pool.
But not in the await chain leading to this entrypoint.

Such code/tasks/continuations could have (expected) externally observable side effects (after we exit the entry point).

But I don't know how to do that at this point. So I guess it will be some next PR.

cc @dicej

Reply:
In the browser we schedule browser callback any time the new job triggers `ThreadPool.RequestWorkerThread()`, but we could not do it here. 

Maybe we need to run `ThreadPoolWorkQueue.Dispatch();` until there was no call to `ThreadPool.RequestWorkerThread()` before we return from the entry point.

cc @SingleAccretion , any thoughts ?

Reply:
I agree that full event loop emulation requires what is effectively:
```cs
bool workItemsLeft;
do
{
    workItemsLeft = ThreadPoolWorkQueue.Dispatch();
}
while (workItemsLeft)
```

I think it's ok to wait for the top-level task only here. It matches "normal" platform's behavior:
```cs
ThreadPool.QueueUserWorkItem(x => { Thread.Sleep(100); Console.WriteLine(); });
> dotnet run
; Nothing printed because thread pool threads are background threads.
```
So it seems unlikely our tests would depend on it.

It _does not_ match Browser's behavior, of course. It is a point to debate whether we'd want the eventual async main experience to match browser or not-browser.